Added support for wide chars via frozen::wstring

marcoffee opened the pull request:
This is an extension of `frozen::string` to support wide strings.
I simply moved the default `frozen::string` to a templated class called `frozen::basic_string` and used it to define `frozen::string` (`frozen::basic_string<char>`) and `frozen::wstring` (`frozen::basic_string<wchar_t>`).
I also added the `_ws` string literal for `frozen::wstring`s.

gracicot commented:
While at it, can you also add unicode string types? I mean `u16string`, `u32string` and maybe also `u8string` if `__cpp_char8_t` is defined.

marcoffee commented:
@gracicot done!

gracicot commented:
@marcoffee just add the literal operator and that'll be good for me! There may also need for unit test, check @serge-sans-paille what need to be added there.

marcoffee commented:
@gracicot string literals added. Just keep in mind that I found out that I can define multiple string literals with the same name, which is resolved at compile time. So I renamed the `frozen::wstring` literal to `_s` and added the unicode string literals also as `_s`.
You simply have to identify your string via [those prefixes](https://en.cppreference.com/w/cpp/language/string_literal).
Regarding the unit test, I will need some ideas, as the existing tests for `frozen::string` pretty much covers the ones for those new strings. Maybe I can focus on adding tests only on the literals.

serge-sans-paille commented:
> Regarding the unit test, I will need some ideas, as the existing tests for `frozen::string` pretty much covers the ones for those new strings. Maybe I can focus on adding tests only on the literals.

 It's probably a good idea to test all string types explicitly (?) in test_str.cpp, isn't it?

serge-sans-paille commented:
oh, and in case that wasn't obvious: thanks for working on this :-)

marcoffee commented:
Sorry for the delay, I got busy working on other things. I will try to add the tests this week.

marcoffee commented:
@serge-sans-paille done!

serge-sans-paille commented:
Thanks a bunch for the contrib and for your patience!
